File: 01_language_comparison/data_structure_manipulation/stack.pl
% Push: Add to front
push(Element, Stack, [Element|Stack]).

% Pop: Remove from front
pop([Top|Rest], Top, Rest).

% Peek: Just read the head
peek([Top|_], Top).

% Example usage:
% ?- push(3, [2,1], S).        => S = [3,2,1].
% ?- pop([3,2,1], X, S2).      => X = 3, S2 = [2,1].
% ?- peek([3,2,1], Top).       => Top = 3.

File: 01_language_comparison/data_structure_manipulation/stack.hs
-- Define a Stack as a list
type Stack a = [a]

-- Push: Add to head
push :: a -> Stack a -> Stack a
push x stack = x : stack

-- Pop: Remove from head
pop :: Stack a -> (a, Stack a)
pop [] = error "Stack Underflow"
pop (x:xs) = (x, xs)

-- Peek: Look at head
peek :: Stack a -> a
peek [] = error "Stack is empty"
peek (x:_) = x

-- Example:
-- let s1 = push 3 []
-- let (top, s2) = pop s1
-- let topVal = peek s2
